client/modules/IDE/components/Preferences/Preferences.unit.test.jsx

```diff
@@ -0,0 +1,830 @@
+import React from 'react';
+import { act } from 'react-dom/test-utils';
+import { fireEvent, render, screen } from '../../../../test-utils';
+import Preferences from './index';
+
+/* props to pass in:
+ * - this.props.fontSize : number
+ * - this.props.autosave : bool
+ * - this.props.autocloseBracketsQuotes : bool
+ * - this.props.linewrap : bool
+ * - this.props.lineNumbers : bool
+ * - this.props.theme : string
+ * - this.props.lintWarning : bool
+ * - this.props.textOutput : bool
+ * - this.props.gridOutput : bool
+ * - this.props.soundOutput : bool
+ * - t from internationalization
+ *
+ * - this.props.setFontSize(fontsize : number)
+ * - this.props.setAutosave(value : bool)
+ * - this.props.setAutocloseBracketsQuotes(value: bool)
+ * - this.props.setLinewrap(value : bool)
+ * - this.props.setLineNumbers(value : bool)
+ * - this.props.setTheme(color : string) -> can be {"light", "dark", "contrast"}
+ * - this.props.setLintWarning(value : bool)
+ * - this.props.setTextOutput(value : bool)
+ * - this.props.setGridOutput(value : bool)
+ * - this.props.setSoundOutput(value : bool)
+ * -
+ */
+
+describe('<Preferences />', () => {
+  let props = {
+    t: jest.fn(),
+    fontSize: 12,
+    autosave: false,
+    autocloseBracketsQuotes: false,
+    linewrap: false,
+    lineNumbers: false,
+    theme: 'contrast',
+    lintWarning: false,
+    textOutput: false,
+    gridOutput: false,
+    soundOutput: false,
+    setFontSize: jest.fn(),
+    setAutosave: jest.fn(),
+    setAutocloseBracketsQuotes: jest.fn(),
+    setLinewrap: jest.fn(),
+    setLineNumbers: jest.fn(),
+    setTheme: jest.fn(),
+    setLintWarning: jest.fn(),
+    setTextOutput: jest.fn(),
+    setGridOutput: jest.fn(),
+    setSoundOutput: jest.fn()
+  };
+
+  const subject = () => render(<Preferences {...props} />);
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('font tests', () => {
+    it('font size increase button says increase', () => {
+      // render the component
+      act(() => {
+        subject();
+      });
+
+      // get ahold of the button for increasing text size
+      const fontPlusButton = screen.getByRole('button', {
+        name: /increase font size/i
+      });
+
+      // check that button says says "Increase"
+      expect(fontPlusButton.textContent.toLowerCase()).toBe('increase');
+    });
+
+    it('increase font size by 2 when clicking plus button', () => {
+      // render the component with font size set to 12
+      act(() => {
+        subject();
+      });
+
+      // get ahold of the button for increasing text size
+      const fontPlusButton = screen.getByRole('button', {
+        name: /increase font size/i
+      });
+
+      // click the button
+      act(() => {
+        fireEvent.click(fontPlusButton);
+      });
+
+      // expect that setFontSize has been called once with the argument 14
+      expect(props.setFontSize).toHaveBeenCalledTimes(1);
```

Comment:
I know we were working on this before we totally got the distinction between unit and integration tests, but this feels like it's testing implementation details, right? Like, what if the names of the Redux functions changed, then the test would break even though the UI hadn't changed at all (i.e. the input and output hadn't changed). Maybe it doesn't even make sense to test this component without Redux because it relies on it so heavily?

Anyway... you totally don't have to change this and maybe it would be good practice for me anyway to try to update it 😄

Reply:
oooo I see what you mean! The input is the user doing something but then the most abstract output that we could test is that the props.setFontSize gets called because nothing on the page actually changes since the parent component hasn't passed in a new set of props in response... I was understanding "implementation detail" as not testing class methods that are hidden within the component, but given a Preferences component ,we can assert that if we provide certain callbacks and then act on the component, the component will call those callbacks with the right values but it doesn't matter what it did in the middle before it finally called the function.

I totally see what you mean! It does make more sense to test it with redux